doc/source/metadata/Working-with-Metadata.rst

```diff
@@ -0,0 +1,295 @@
+  a step in the suite
+* When and how metadata treatment should be changed when code is modified
+
+Principles of objects and metadata
+----------------------------------
+
+In object-oriented programming, metadata is intrinsically linked to the data.
+Any publicly callable routine (function or public class method)
+that updates the data inside a cube must also update any appropriate metadata,
+so that the object returned is correct and self-consistent.
+
+For example, the threshold plugin converts the data in a cube into probabilities.
+Metadata wise, it must also do the following:
+
+* Produce a complete and correct threshold-type coordinate describing
+  the threshold values and the nature of the data relative to the threshold
+  (``greater than``, ``less than``, etc)
+* Update the name and units of the cube to reflect that it now contains
+  probabilities (e.g. ``probability_of_X_above_threshold``)
+* Update any cell methods to be consistent with thresholded data
+
+Encapsulation / responsibility
+******************************
+
+Public interfaces **must** take full responsibility for the objects they act upon.
+Any function or method that does not update **all** of the required metadata
+should either:
+
+* Operate only on the data matrix
+  (i.e. take ``cube.data`` rather than ``cube`` as an argument), or
+* If within a plugin, be a private method (not callable from outside the plugin)
+
+Extending the example above: any method within the threshold plugin that returns
+a cube with thresholded data, but without a new probability name, should be
+a private method.
+
+The corollary of this is that functions or classes **must not** modify
+data or metadata that is outside of their scope.
+A plugin or function's purpose therefore defines precisely which metadata
+it must update: no less, and no more.
+
+Writing new functions or plugins
+--------------------------------
+
+The code in IMPROVER generally falls into one of the following categories:
+
+Post-processing code
+    This modifies an existing parameter, for example through
+    spatial post-processing
+    (neighbourhood processing or recursive filtering),
+    or converts between different
+    representations of the probability distribution
+    (realizations, probabilities and percentiles).
+    The name of the underlying parameter (e.g. air_temperature) remains the same.
+
+New parameter creation
+    This takes various inputs to create a new parameter,
+    for example lapse rate or weather symbols.
+    A new parameter name is required.
+
+Calculation of post-processing parameters
+    For example EMOS coefficients or reliability tables.
+    The output is generally not a data cube, but a cube of parameters
+    that will be used to post-process the data at a later stage.
+    Plugins to apply these parameters would be post-processing code.
+
+**General utility code**
+    .. empty description, so the bold it make it consistent
```

Comment:
I'm not sure the bold was sufficient to leave me unconfused, I just thought the test below had not been indented correctly. Maybe you could put some brief content in here, like a couple of very simple examples?

Reply:
Fair point - I've added a filler sentence and also adapted the lead-in sentence to make it explicit that there are 4 categories.